File: platform/ios/PolyPodApp/PodApi/Triplestore/Triplestore.swift
import Foundation
import MessagePack
import PolyPodCoreSwift

protocol Triplestore {
    func query(
        query: String,
        completionHandler: @escaping (MessagePackValue, Error?) -> Void
    )
    func update(
        query: String,
        completionHandler: @escaping (Error?) -> Void
    )
}

extension Core: Triplestore {
    func query(
        query: String,
        completionHandler: @escaping (MessagePackValue, Error?) -> Void
    ) {
        let result: Result<MessagePackValue, Error> = executeRequest(.executeRdfQuery(args: query))
        switch result {
        case .success(let result):
            completionHandler(result, nil)
        case .failure(let error):
            completionHandler(nil, error)
        }
    }
    
    func update(
        query: String,
        completionHandler: @escaping (Error?) -> Void
    ) {
        switch executeRequest(.executeRdfUpdate(args: query)) {
        case .success:
            completionHandler(nil)
        case .failure(let error):
            completionHandler(error)
        }
    }
}

File: platform/android/app/src/main/kotlin/coop/polypoly/polypod/api/triplestore/Triplestore.kt
package coop.polypoly.polypod.api.triplestore

import coop.polypoly.core.Core
import coop.polypoly.core.CoreRequest
import org.msgpack.value.Value

open class Triplestore() {
    open fun query(query: String): Value {
        return Core.executeRequest(CoreRequest.ExecuteRdfQuery(query)) { it }
    }

    open fun update(query: String) {
        return Core.executeRequest(CoreRequest.ExecuteRdfUpdate(query))
    }
}
